packages/dds/tree/src/shared-tree-core/editManager.ts

```diff
@@ -656,66 +664,73 @@ export class EditManager<
-					};
+		for (const newCommit of newCommits) {
+			assert(
+				sequenceNumber >= // This is ">=", not ">" because changes in the same batch will have the same sequence number
+					(this.sequenceMap.maxKey()?.sequenceNumber ?? minimumPossibleSequenceNumber),
+				0xa64 /* Attempted to sequence change with an outdated sequence number */,
+			);
 
-		if (newCommit.sessionId === this.localSessionId) {
-			return this.fastForwardNextLocalCommit(sequenceId);
-		}
+			const commitsSequenceNumber = this.getBatch(sequenceNumber);
+			const sequenceId: SequenceId =
+				commitsSequenceNumber.length === 0
+					? {
+							sequenceNumber,
+						}
+					: {
+							sequenceNumber,
+							indexInBatch: commitsSequenceNumber.length,
+						};
+
+			if (newCommit.sessionId === this.localSessionId) {
+				this.fastForwardNextLocalCommit(sequenceId);
+				continue;
+			}
 
-		// Get the revision that the remote change is based on
-		const [, baseRevisionInTrunk] = this.getClosestTrunkCommit(referenceSequenceNumber);
-		// Rebase that branch over the part of the trunk up to the base revision
-		// This will be a no-op if the sending client has not advanced since the last time we received an edit from it
-		const peerLocalBranch = getOrCreate(
-			this.peerLocalBranches,
-			newCommit.sessionId,
-			() => new SharedTreeBranch(baseRevisionInTrunk, this.changeFamily, this.mintRevisionTag),
-		);
-		peerLocalBranch.rebaseOnto(this.trunk, baseRevisionInTrunk);
-
-		if (peerLocalBranch.getHead() === this.trunk.getHead()) {
-			// If the branch is fully caught up and empty after being rebased, then push to the trunk directly
-			this.pushCommitToTrunk(sequenceId, newCommit);
-			peerLocalBranch.setHead(this.trunk.getHead());
-		} else {
-			// Otherwise, rebase the change over the trunk and append it, and append the original change to the peer branch.
-			const { duration, output: newChangeFullyRebased } = measure(() =>
-				rebaseChange(
-					this.changeFamily.rebaser,
-					newCommit,
-					peerLocalBranch.getHead(),
-					this.trunk.getHead(),
-					this.mintRevisionTag,
-				),
+			// Step 1 - Recreate the peer remote client's local environment.
+			// Get the revision that the remote change is based on
+			const [, baseRevisionInTrunk] = this.getClosestTrunkCommit(referenceSequenceNumber);
+			// Rebase that peer local branch over the part of the trunk up to the base revision
+			// This will be a no-op if the sending client has not advanced since the last time we received an edit from it
+			const peerLocalBranch = getOrCreate(
+				this.peerLocalBranches,
+				newCommit.sessionId,
+				() =>
+					new SharedTreeBranch(baseRevisionInTrunk, this.changeFamily, this.mintRevisionTag),
 			);
+			peerLocalBranch.rebaseOnto(this.trunk, baseRevisionInTrunk);
 
-			this.telemetryEventBatcher?.accumulateAndLog({
-				duration,
-				...newChangeFullyRebased.telemetryProperties,
-			});
+			// Step 2 - Rebase the change over the trunk and append it to the peer branch
```

Comment:
This might sound like we're appending the rebased version of the change to the peer branch.
```suggestion
			// Step 2 - Append the change to the peer branch. Rebase the change to the tip of the trunk.
```
It's a bit unfortunate that the code is doing things in the opposite order. Maybe we can fix that in your next PR since it'll be changing that code.